File: orgarif-client/src/domain/command.ts
import { LoginResult, RegisterResult, UserInfos } from '../domain/user';
import {
  DeliberationId,
  EluId,
  InstanceId,
  LienDeliberationId,
  NatureJuridiqueId,
  OrganismeId,
  RepresentantId,
  SecteurId,
  TypeStructureId,
} from './id';
import { RepresentantOrSuppleant } from './organisme';
import { LocalDate } from './time';

export interface AddInstanceCommand {
  nomInstance: string;
  organismeId: OrganismeId;
}

export interface AddInstanceCommandResponse {
  id: InstanceId;
}

export interface AddLienDeliberationCommand {
  deliberationId: DeliberationId;
  organismeId: OrganismeId;
  instanceId?: InstanceId;
}

export interface AddLienDeliberationCommandResponse {
  lienDeliberationId: LienDeliberationId;
}

export interface AddRepresentantCommand {
  eluId: EluId;
  organismeId: OrganismeId;
  instanceId?: InstanceId;
  representantOrSuppleant: RepresentantOrSuppleant;
}

export interface AddRepresentantCommandResponse {
  id: RepresentantId;
}

export interface CreateDeliberationAndAddLienCommand {
  libelle: string;
  deliberationDate: LocalDate;
  organismeId: OrganismeId;
  instanceId?: InstanceId;
}

export interface CreateDeliberationAndAddLienCommandResponse {
  deliberationId: DeliberationId;
  lienDeliberationId: LienDeliberationId;
}

export interface CreateOrganismeCommand {
  nom: string;
}

export interface CreateOrganismeCommandResponse {
  id: OrganismeId;
}

export interface DeleteInstanceCommand {
  id: InstanceId;
}

export interface DeleteRepresentantCommand {
  id: RepresentantId;
}

export interface LoginCommand {
  login: string;
  password: string;
}

export interface LoginCommandResponse {
  result: LoginResult;
  userinfos?: UserInfos;
}

export interface MoveRepresentantCommand {
  id: RepresentantId;
  toOrganismeId: OrganismeId;
  toInstanceId?: InstanceId;
  toPosition: number;
  toRepresentantOrSuppleant: RepresentantOrSuppleant;
}

export interface RegisterCommand {
  mail: string;
  password: string;
}

export interface RegisterCommandResponse {
  result: RegisterResult;
  userinfos?: UserInfos;
}

export interface UpdateOrganismeNatureJuridiqueCommand {
  id: OrganismeId;
  natureJuridiqueId: NatureJuridiqueId;
}

export interface UpdateOrganismeSecteurCommand {
  id: OrganismeId;
  secteurId: SecteurId;
}

export interface UpdateOrganismeTypeStructureCommand {
  id: OrganismeId;
  typeStructureId: TypeStructureId;
}

File: orgarif-server/src/main/kotlin/orgarif/command/command.kt
package orgarif.command

import RepresentantOrSuppleant
import com.fasterxml.jackson.annotation.JsonTypeInfo
import orgarif.domain.*
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@JsonTypeInfo(use = JsonTypeInfo.Id.MINIMAL_CLASS, include = JsonTypeInfo.As.PROPERTY, property = "objectType")
sealed class Command

sealed class CommandResponse

object EmptyCommandResponse : CommandResponse()

interface CommandHandler<C : Command, R : CommandResponse> {
    fun handle(command: C, userSession: UserSession?, request: HttpServletRequest, response: HttpServletResponse): R
}

abstract class LoggedOutServletCommandHandler<C : Command, R : CommandResponse> : CommandHandler<C, R> {
    override fun handle(command: C, userSession: UserSession?, request: HttpServletRequest, response: HttpServletResponse): R {
        if (userSession != null) {
            throw RuntimeException()
        }
        return handle(command, request, response)
    }

    abstract fun handle(command: C, request: HttpServletRequest, response: HttpServletResponse): R
}

//abstract class UserSessionCommandHandler<C : Command, R : CommandResponse> : CommandHandler<C, R> {
//    override fun doHandle(command: C, userSession: UserSession?, request: HttpServletRequest, response: HttpServletResponse): R {
//        return handle(command, userSession)
//    }
//
//    abstract fun handle(command: C, userSession: UserSession?): R
//}

abstract class LoggedOutCommandHandler<C : Command, R : CommandResponse> : CommandHandler<C, R> {
    override fun handle(command: C, userSession: UserSession?, request: HttpServletRequest, response: HttpServletResponse): R {
        if (userSession != null) {
            throw RuntimeException()
        }
        return handle(command)
    }

    abstract fun handle(command: C): R
}

abstract class LoggedInCommandHandler<C : Command, R : CommandResponse> : CommandHandler<C, R> {
    override fun handle(command: C, userSession: UserSession?, request: HttpServletRequest, response: HttpServletResponse): R {
        if (userSession == null) {
            throw RuntimeException()
        }
        return handle(command, userSession)
    }

    abstract fun handle(command: C, userSession: UserSession): R
}

abstract class NeutralCommandHandler<C : Command, R : CommandResponse> : CommandHandler<C, R> {
    override fun handle(command: C, userSession: UserSession?, request: HttpServletRequest, response: HttpServletResponse): R {
        return handle(command)
    }

    abstract fun handle(command: C): R
}

data class CreateOrganismeCommand(val nom: String) : Command()

data class CreateOrganismeCommandResponse(val id: OrganismeId) : CommandResponse()

// [doc] login as username|mail
data class LoginCommand(val login: String,
                        val password: PlainStringPassword) : Command()

data class LoginCommandResponse(val result: LoginResult,
                                val userinfos: UserInfos?,
                                val csrfToken: String?) : CommandResponse()

data class MoveRepresentantCommand(val id: RepresentantId,
                                   val toOrganismeId: OrganismeId,
                                   val toInstanceId: InstanceId?,
                                   val toPosition: Int,
                                   val toRepresentantOrSuppleant: RepresentantOrSuppleant) : Command()

data class RegisterCommand(val mail: String,
                           val password: PlainStringPassword,
                           val displayName: String) : Command()

data class RegisterCommandResponse(val result: RegisterResult,
                                   val userinfos: UserInfos?) : CommandResponse()

data class UpdateOrganismeNatureJuridiqueCommand(val id: OrganismeId,
                                                 val natureJuridiqueId: NatureJuridiqueId) : Command()

data class UpdateOrganismeSecteurCommand(val id: OrganismeId,
                                         val secteurId: SecteurId) : Command()

data class UpdateOrganismeTypeStructureCommand(val id: OrganismeId,
                                               val typeStructureId: TypeStructureId) : Command()